In [1]:
# Import required libraries

import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [2]:
# Load raw dataset

data_path = "../data/raw/credit_card_customers.csv"

df = pd.read_csv(data_path)

df.head()

,CLIENTNUM,Attrition_Flag,Customer_Age,Gender,Dependent_count,Education_Level,Marital_Status,Income_Category,Card_Category,Months_on_book,...,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt,Total_Trans_Ct,Total_Ct_Chng_Q4_Q1,Avg_Utilization_Ratio,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2
0,768805383,Existing Customer,45,M,3,High School,Married,$60K - $80K,Blue,39,...,12691.0,777,11914.0,1.335,1144,42,1.625,0.061,0.000093,0.99991
1,818770008,Existing Customer,49,F,5,Graduate,Single,Less than $40K,Blue,44,...,8256.0,864,7392.0,1.541,1291,33,3.714,0.105,0.000057,0.99994
2,713982108,Existing Customer,51,M,3,Graduate,Married,$80K - $120K,Blue,36,...,3418.0,0,3418.0,2.594,1887,20,2.333,0.000,0.000021,0.99998
3,769911858,Existing Customer,40,F,4,High School,Unknown,Less than $40K,Blue,34,...,3313.0,2517,796.0,1.405,1171,20,2.333,0.760,0.000134,0.99987
4,709106358,Existing Customer,40,M,3,Uneducated,Married,$60K - $80K,Blue,21,...,4716.0,0,4716.0,2.175,816,28,2.500,0.000,0.000022,0.99998


In [3]:
# Check dataset shape

df.shape

(10127, 23)

In [4]:
# Check missing values

missing_values = df.isnull().sum().sort_values(ascending=False)

missing_values

CLIENTNUM                                                                                                                             0
Contacts_Count_12_mon                                                                                                                 0
Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1    0
Avg_Utilization_Ratio                                                                                                                 0
Total_Ct_Chng_Q4_Q1                                                                                                                   0
Total_Trans_Ct                                                                                                                        0
Total_Trans_Amt                                                                                                                       0
Total_Amt_Chng_Q4_Q1                            

In [5]:
# Check missing value percentage

missing_percentage = (df.isnull().mean() * 100).sort_values(ascending=False)

missing_percentage

CLIENTNUM                                                                                                                             0.0
Contacts_Count_12_mon                                                                                                                 0.0
Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1    0.0
Avg_Utilization_Ratio                                                                                                                 0.0
Total_Ct_Chng_Q4_Q1                                                                                                                   0.0
Total_Trans_Ct                                                                                                                        0.0
Total_Trans_Amt                                                                                                                       0.0
Total_Amt_Chng_Q4_Q1              

In [6]:
# Check hidden missing-like values in categorical columns

categorical_features = [
    "Gender",
    "Education_Level",
    "Marital_Status",
    "Income_Category",
    "Card_Category"
]

for col in categorical_features:
    print(f"\n{col}")
    print(df[col].value_counts())


Gender
Gender
F    5358
M    4769
Name: count, dtype: int64

Education_Level
Education_Level
Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: count, dtype: int64

Marital_Status
Marital_Status
Married     4687
Single      3943
Unknown      749
Divorced     748
Name: count, dtype: int64

Income_Category
Income_Category
Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: count, dtype: int64

Card_Category
Card_Category
Blue        9436
Silver       555
Gold         116
Platinum      20
Name: count, dtype: int64


In [7]:
df.isnull().sum()

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

In [8]:
categorical_features = [
    "Gender",
    "Education_Level",
    "Marital_Status",
    "Income_Category",
    "Card_Category"
]

for col in categorical_features:
    print(f"\n{'='*50}")
    print(col)
    print(df[col].value_counts())


Gender
Gender
F    5358
M    4769
Name: count, dtype: int64

Education_Level
Education_Level
Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: count, dtype: int64

Marital_Status
Marital_Status
Married     4687
Single      3943
Unknown      749
Divorced     748
Name: count, dtype: int64

Income_Category
Income_Category
Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: count, dtype: int64

Card_Category
Card_Category
Blue        9436
Silver       555
Gold         116
Platinum      20
Name: count, dtype: int64


## 3.1 Missing Value Treatment – Hidden Missing Category Analysis

Although the dataset contained **no true missing values (`NaN`)**, an additional investigation was performed to identify **hidden missing categories** within categorical variables.

The following categorical features were examined:

- Gender
- Education_Level
- Marital_Status
- Income_Category
- Card_Category

### Findings

#### Education_Level

The `Education_Level` feature contained an `"Unknown"` category.

| Category | Count |
|-----------|--------|
| Graduate | 3128 |
| High School | 2013 |
| Unknown | 1519 |
| Uneducated | 1487 |
| College | 1013 |
| Post-Graduate | 516 |
| Doctorate | 451 |

---

#### Marital_Status

The `Marital_Status` feature also contained an `"Unknown"` category.

| Category | Count |
|-----------|--------|
| Married | 4687 |
| Single | 3943 |
| Unknown | 749 |
| Divorced | 748 |

---

#### Income_Category

The `Income_Category` feature contained an `"Unknown"` category as well.

| Category | Count |
|-----------|--------|
| Less than $40K | 3561 |
| $40K - $60K | 1790 |
| $80K - $120K | 1535 |
| $60K - $80K | 1402 |
| Unknown | 1112 |
| $120K + | 727 |

---

### Interpretation

The presence of `"Unknown"` values does not represent true missing data from a technical perspective. Instead, these categories indicate that the corresponding customer information was either unavailable or undisclosed.

In banking and customer analytics applications, such categories may themselves carry predictive information. For example, customers who choose not to disclose their income or educational background may exhibit behavioral patterns different from those who provide complete information.

Therefore, treating these observations as ordinary missing values and imputing or removing them could lead to the loss of potentially valuable signals.

------

### Decision

The `"Unknown"` categories will be **retained as valid categories** during preprocessing.

No imputation or row removal will be performed for these observations.

### Final Outcome of Missing Value Treatment

| Component | Decision |
|------------|-----------|
| True Missing Values (`NaN`) | No treatment required |
| Numerical Imputation | Not required |
| Categorical Imputation | Not required |
| Row Deletion | Not performed |
| `"Unknown"` Categories | Retained as valid categories |

### Business Justification

Preserving `"Unknown"` categories enables the model to learn whether the absence of certain customer information is associated with customer churn. This approach aligns with industry practices commonly adopted in financial and customer retention modeling.

In [9]:
numerical_features = [
    "Customer_Age",
    "Dependent_count",
    "Months_on_book",
    "Total_Relationship_Count",
    "Months_Inactive_12_mon",
    "Contacts_Count_12_mon",
    "Credit_Limit",
    "Total_Revolving_Bal",
    "Avg_Open_To_Buy",
    "Total_Amt_Chng_Q4_Q1",
    "Total_Trans_Amt",
    "Total_Trans_Ct",
    "Total_Ct_Chng_Q4_Q1",
    "Avg_Utilization_Ratio"
]

print(f"Number of numerical features: {len(numerical_features)}")
numerical_features

Number of numerical features: 14


['Customer_Age',
 'Dependent_count',
 'Months_on_book',
 'Total_Relationship_Count',
 'Months_Inactive_12_mon',
 'Contacts_Count_12_mon',
 'Credit_Limit',
 'Total_Revolving_Bal',
 'Avg_Open_To_Buy',
 'Total_Amt_Chng_Q4_Q1',
 'Total_Trans_Amt',
 'Total_Trans_Ct',
 'Total_Ct_Chng_Q4_Q1',
 'Avg_Utilization_Ratio']

In [10]:
df[numerical_features].describe().T

,count,mean,std,min,25%,50%,75%,max
Customer_Age,10127.0,46.325960,8.016814,26.0,41.000,46.000,52.000,73.000
Dependent_count,10127.0,2.346203,1.298908,0.0,1.000,2.000,3.000,5.000
Months_on_book,10127.0,35.928409,7.986416,13.0,31.000,36.000,40.000,56.000
Total_Relationship_Count,10127.0,3.812580,1.554408,1.0,3.000,4.000,5.000,6.000
Months_Inactive_12_mon,10127.0,2.341167,1.010622,0.0,2.000,2.000,3.000,6.000
Contacts_Count_12_mon,10127.0,2.455317,1.106225,0.0,2.000,2.000,3.000,6.000
Credit_Limit,10127.0,8631.953698,9088.776650,1438.3,2555.000,4549.000,11067.500,34516.000
Total_Revolving_Bal,10127.0,1162.814061,814.987335,0.0,359.000,1276.000,1784.000,2517.000
Avg_Open_To_Buy,10127.0,7469.139637,9090.685324,3.0,1324.500,3474.000,9859.000,34516.000
Total_Amt_Chng_Q4_Q1,10127.0,0.759941,0.219207,0.0,0.631,0.736,0.859,3.397


In [11]:
categorical_features = [
    "Gender",
    "Education_Level",
    "Marital_Status",
    "Income_Category",
    "Card_Category"
]

categorical_features

['Gender',
 'Education_Level',
 'Marital_Status',
 'Income_Category',
 'Card_Category']

In [12]:
df["Attrition_Flag"] = df["Attrition_Flag"].map({
    "Existing Customer": 0,
    "Attrited Customer": 1
})

In [13]:
df["Attrition_Flag"].value_counts()

Attrition_Flag
0    8500
1    1627
Name: count, dtype: int64

In [14]:
columns_to_drop = [
    "CLIENTNUM",
    "Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1",
    "Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2"
]

df = df.drop(columns=columns_to_drop)

In [15]:
df.shape
df.columns

Index(['Attrition_Flag', 'Customer_Age', 'Gender', 'Dependent_count',
       'Education_Level', 'Marital_Status', 'Income_Category', 'Card_Category',
       'Months_on_book', 'Total_Relationship_Count', 'Months_Inactive_12_mon',
       'Contacts_Count_12_mon', 'Credit_Limit', 'Total_Revolving_Bal',
       'Avg_Open_To_Buy', 'Total_Amt_Chng_Q4_Q1', 'Total_Trans_Amt',
       'Total_Trans_Ct', 'Total_Ct_Chng_Q4_Q1', 'Avg_Utilization_Ratio'],
      dtype='str')

In [16]:
X = df.drop("Attrition_Flag", axis=1)

y = df["Attrition_Flag"]

In [17]:
print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (10127, 19)
y shape: (10127,)


In [18]:
y.value_counts()

Attrition_Flag
0    8500
1    1627
Name: count, dtype: int64

In [19]:
y.value_counts(normalize=True) * 100

Attrition_Flag
0    83.934038
1    16.065962
Name: proportion, dtype: float64

In [20]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

In [21]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

print("\nTraining target distribution:")
print(y_train.value_counts(normalize=True) * 100)

print("\nTesting target distribution:")
print(y_test.value_counts(normalize=True) * 100)

X_train shape: (8101, 19)
X_test shape: (2026, 19)

Training target distribution:
Attrition_Flag
0    83.92791
1    16.07209
Name: proportion, dtype: float64

Testing target distribution:
Attrition_Flag
0    83.958539
1    16.041461
Name: proportion, dtype: float64


In [22]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [23]:
numerical_features = [
    "Customer_Age",
    "Dependent_count",
    "Months_on_book",
    "Total_Relationship_Count",
    "Months_Inactive_12_mon",
    "Contacts_Count_12_mon",
    "Credit_Limit",
    "Total_Revolving_Bal",
    "Avg_Open_To_Buy",
    "Total_Amt_Chng_Q4_Q1",
    "Total_Trans_Amt",
    "Total_Trans_Ct",
    "Total_Ct_Chng_Q4_Q1",
    "Avg_Utilization_Ratio"
]

categorical_features = [
    "Gender",
    "Education_Level",
    "Marital_Status",
    "Income_Category",
    "Card_Category"
]

In [24]:
numeric_transformer = Pipeline(steps=[
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

In [25]:
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

### Now, let's just fit the preprocessor only on training data.

In [26]:
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

### fit_transform(X_train):
 learns scaling + category mapping from training data only

### transform(X_test):
 applies the same learned rules to unseen test data

In [27]:
print("X_train_processed shape:", X_train_processed.shape)
print("X_test_processed shape:", X_test_processed.shape)

X_train_processed shape: (8101, 37)
X_test_processed shape: (2026, 37)


In [ ]:
# Get final feature names after preprocessing

feature_names = preprocessor.get_feature_names_out()

feature_names

array(['num__Customer_Age', 'num__Dependent_count', 'num__Months_on_book',
       'num__Total_Relationship_Count', 'num__Months_Inactive_12_mon',
       'num__Contacts_Count_12_mon', 'num__Credit_Limit',
       'num__Total_Revolving_Bal', 'num__Avg_Open_To_Buy',
       'num__Total_Amt_Chng_Q4_Q1', 'num__Total_Trans_Amt',
       'num__Total_Trans_Ct', 'num__Total_Ct_Chng_Q4_Q1',
       'num__Avg_Utilization_Ratio', 'cat__Gender_F', 'cat__Gender_M',
       'cat__Education_Level_College', 'cat__Education_Level_Doctorate',
       'cat__Education_Level_Graduate',
       'cat__Education_Level_High School',
       'cat__Education_Level_Post-Graduate',
       'cat__Education_Level_Uneducated', 'cat__Education_Level_Unknown',
       'cat__Marital_Status_Divorced', 'cat__Marital_Status_Married',
       'cat__Marital_Status_Single', 'cat__Marital_Status_Unknown',
       'cat__Income_Category_$120K +', 'cat__Income_Category_$40K - $60K',
       'cat__Income_Category_$60K - $80K',
       'cat__Inco

In [29]:
X_train_processed_df = pd.DataFrame(
    X_train_processed,
    columns=feature_names,
    index=X_train.index
)

X_test_processed_df = pd.DataFrame(
    X_test_processed,
    columns=feature_names,
    index=X_test.index
)

X_train_processed_df.head()

,num__Customer_Age,num__Dependent_count,num__Months_on_book,num__Total_Relationship_Count,num__Months_Inactive_12_mon,num__Contacts_Count_12_mon,num__Credit_Limit,num__Total_Revolving_Bal,num__Avg_Open_To_Buy,num__Total_Amt_Chng_Q4_Q1,...,cat__Income_Category_$120K +,cat__Income_Category_$40K - $60K,cat__Income_Category_$60K - $80K,cat__Income_Category_$80K - $120K,cat__Income_Category_Less than $40K,cat__Income_Category_Unknown,cat__Card_Category_Blue,cat__Card_Category_Gold,cat__Card_Category_Platinum,cat__Card_Category_Silver
2856,-1.301232,-1.809900,-1.513205,0.122196,0.646495,-0.411236,-0.665434,1.303942,-0.783057,-0.335597,...,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0
6515,-0.299340,0.501503,-0.000700,0.765411,-0.342413,0.489635,1.848571,-1.433246,1.977862,0.407489,...,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
7141,-0.048867,-0.268965,-0.756952,1.408625,0.646495,-1.312106,0.342117,-0.310139,0.370094,0.603518,...,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0
632,-1.301232,-0.268965,-1.513205,-0.521019,-1.331320,0.489635,-0.604095,0.522064,-0.651190,0.498665,...,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0
3496,0.452079,-1.039432,0.503468,0.122196,-0.342413,0.489635,2.871622,0.021269,2.869713,-0.157803,...,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0


In [30]:
import os
import joblib

os.makedirs("../data/processed", exist_ok=True)
os.makedirs("../src/models", exist_ok=True)

In [31]:
X_train_processed_df.to_csv("../data/processed/X_train_processed.csv", index=False)
X_test_processed_df.to_csv("../data/processed/X_test_processed.csv", index=False)

y_train.to_csv("../data/processed/y_train.csv", index=False)
y_test.to_csv("../data/processed/y_test.csv", index=False)

In [32]:
joblib.dump(preprocessor, "../src/models/preprocessor.pkl")

['../src/models/preprocessor.pkl']

In [33]:
print("Saved processed datasets and preprocessor successfully.")
print("X_train:", X_train_processed_df.shape)
print("X_test:", X_test_processed_df.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

Saved processed datasets and preprocessor successfully.
X_train: (8101, 37)
X_test: (2026, 37)
y_train: (8101,)
y_test: (2026,)


In [34]:
X_train = pd.read_csv("../data/processed/X_train_processed.csv")
X_test = pd.read_csv("../data/processed/X_test_processed.csv")
y_train = pd.read_csv("../data/processed/y_train.csv").squeeze()
y_test = pd.read_csv("../data/processed/y_test.csv").squeeze()

In [35]:
import os

print(os.path.exists("../data/processed/X_train_processed.csv"))
print(os.path.exists("../data/processed/X_test_processed.csv"))
print(os.path.exists("../data/processed/y_train.csv"))
print(os.path.exists("../data/processed/y_test.csv"))
print(os.path.exists("../src/models/preprocessor.pkl"))

True
True
True
True
True
